internal/konnect/node_agent_test.go

```diff
@@ -381,3 +367,99 @@ func TestNodeAgent_StartDoesntReturnUntilContextGetsCancelled(t *testing.T) {
+	nodeClient := newMockNodeClient(nil)
+	configStatusQueue := newMockConfigStatusNotifier()
+	gatewayClientsChangesNotifier := newMockGatewayClientsNotifier()
+
+	nodeAgent := konnect.NewNodeAgent(
+		testHostname,
+		testKicVersion,
+		konnect.DefaultRefreshNodePeriod,
+		logr.Discard(),
+		nodeClient,
+		configStatusQueue,
+		newMockGatewayInstanceGetter(nil),
+		gatewayClientsChangesNotifier,
+		newMockManagerInstanceIDProvider(uuid.New()),
+	)
+
+	runAgent(t, nodeAgent)
+
+	testCases := []struct {
+		notifiedConfigStatus    clients.ConfigStatus
+		expectedControllerState nodes.IngressControllerState
+	}{
+		{
+			notifiedConfigStatus:    clients.ConfigStatusOK,
+			expectedControllerState: nodes.IngressControllerStateOperational,
+		},
+		{
+			notifiedConfigStatus:    clients.ConfigStatusTranslationErrorHappened,
+			expectedControllerState: nodes.IngressControllerStatePartialConfigFail,
+		},
+		{
+			notifiedConfigStatus:    clients.ConfigStatusApplyFailed,
+			expectedControllerState: nodes.IngressControllerStateInoperable,
+		},
+		{
+			notifiedConfigStatus:    clients.ConfigStatusOKKonnectApplyFailed,
+			expectedControllerState: nodes.IngressControllerStateOperationalKonnectOutOfSync,
+		},
+		{
+			notifiedConfigStatus:    clients.ConfigStatusTranslationErrorHappenedKonnectApplyFailed,
+			expectedControllerState: nodes.IngressControllerStatePartialConfigFailKonnectOutOfSync,
+		},
+		{
+			notifiedConfigStatus:    clients.ConfigStatusApplyFailedKonnectApplyFailed,
+			expectedControllerState: nodes.IngressControllerStateInoperableKonnectOutOfSync,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(fmt.Sprint(tc.notifiedConfigStatus), func(t *testing.T) {
+			configStatusQueue.Notify(tc.notifiedConfigStatus)
+
+			require.Eventually(t, func() bool {
+				controllerNode, ok := lo.Find(nodeClient.MustAllNodes(), func(n *nodes.NodeItem) bool {
+					return n.Type == nodes.NodeTypeIngressController
+				})
+				if !ok {
+					t.Log("controller node not found")
+					return false
+				}
+
+				if controllerNode.Status != string(tc.expectedControllerState) {
+					t.Logf("expected controller node status to be %q, got %q", tc.expectedControllerState, controllerNode.Status)
+					return false
+				}
+
+				return true
+			}, time.Second, time.Millisecond)
+		})
+	}
+}
+
+// runAgent runs the agent in a goroutine and cancels the context after the test is done, ensuring that the agent
+// doesn't return prematurely.
+func runAgent(t *testing.T, nodeAgent *konnect.NodeAgent) {
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// To be used as a barrier to ensure that the agent returned after the context was cancelled.
+	agentReturned := make(chan struct{})
+	go func() {
+		err := nodeAgent.Start(ctx)
+		require.NoError(t, err, "expected no error even when the context is cancelled")
+		close(agentReturned)
+	}()
+
+	t.Cleanup(func() {
+		cancel()
```

Comment:
I wanted to comment that this feels a bit too complex for what it's trying to achieve but then I realized that this is the day of signalling controller-runtime's runnable managed by the controller, i.e. just using the provided context.

The fact I don't like it, is that the logs (or [code](https://github.com/Kong/kubernetes-ingress-controller/blob/1dd07258f3169020935cdaed6b556c7c12fc130d/internal/konnect/node_agent.go#L165-L166)) might indicate that there was a failure were in fact the context was cancelled.

I'm not trying to suggest any actions on this, just stating my observations.